### PR TITLE
Nv 1205 weak password policy

### PR DESCRIPTION
**Pull request by BiswaViraj**
### What change does this PR introduce? 
- Adds strong password policy for user registration and reset password
- Adds a password strength meter on the UI
 

### Other information (Screenshots)

https://user-images.githubusercontent.com/29251776/207357484-1ab71fb1-adc5-449e-bb81-6051c6a7c96a.mov

image

### CODE REVIEW

File: apps/api/src/app/auth/auth.controller.ts [hunk 1]
```diff
@@ -147,7 +148,7 @@ export class AuthController {
   }
 
   @Post('/reset')
-  async passwordReset(@Body() body: { password: string; token: string }) {
+  async passwordReset(@Body() body: PasswordResetBodyDto) {
```
Comment (scopsy) [inline apps/api/src/app/auth/auth.controller.ts:151]: Good catch!

File: apps/api/src/app/auth/e2e/user-registration.e2e.ts [hunk 2]
```diff
@@ -96,7 +96,7 @@ describe('User registration - /auth/register (POST)', async () => {
       email: 'Testy.test@gmail.com',
       firstName: 'Test',
       lastName: 'User',
-      password: '123456789',
+      password: '123@Qwerty',
```
Comment (scopsy) [inline apps/api/src/app/auth/e2e/user-registration.e2e.ts:99]: Do we have a test today to validate the password strength logic? Maybe we can create a small case for it here 🙏

**Review by scopsy — APPROVED** [review]
Looks good, please take a look at my note on tests 🎉

File: libs/shared/src/consts/password-helper/passwordHelper.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+export const passwordConstraints = {
+  minLength: 8,
+  maxLength: 64,
+  pattern: /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{8,64}$/,
```
Comment (p-fernandez) [inline libs/shared/src/consts/password-helper/passwordHelper.ts:4]: Seeing the RegExp I think we should inform the users which are the valid special characters. Some of them are missing and users can receive an error message and be confused about it.

Comment (BiswaViraj) [reply libs/shared/src/consts/password-helper/passwordHelper.ts:4]: You're right, will add them to the error message

File: apps/api/src/app/auth/e2e/user-registration.e2e.ts (hunk 2 above)
Comment (BiswaViraj) [reply apps/api/src/app/auth/e2e/user-registration.e2e.ts:99]: Couldn't find any exisitng tests,
will add them

**Comment by scopsy** [comment]
@BiswaViraj now seems like tests are passing, if you will have a chance to take a look at codeQL suggestion to improve this regex 🙏

File: libs/shared/src/consts/password-helper/passwordHelper.ts (hunk 3 above)
Comment (ainouzgali) [reply libs/shared/src/consts/password-helper/passwordHelper.ts]: There is this example in [Mantine to show password strength ](https://mantine.dev/core/password-input/#strength-meter-example)
I kind of like that it shows requirements separately. Let me know if its any help @BiswaViraj  :)
Screen Shot 2022-12-18 at 13 37 09

Comment (BiswaViraj) [reply libs/shared/src/consts/password-helper/passwordHelper.ts]: Yeah I think we can use a similar design will just replace the bar with the strength meter
Thanks @ainouzgali

**Comment by BiswaViraj** [comment]
https://user-images.githubusercontent.com/29251776/208458454-3e816a08-add1-4a81-ada1-183193ddaf7e.mov

**Review by scopsy — APPROVED** [review]

**Pull request merged**